File: src/main/java/com/gmail/samueler53/fwhumanstratego/gui/GamesGui.java
package com.gmail.samueler53.fwhumanstratego.gui;

import com.github.stefvanschie.inventoryframework.Gui;
import com.github.stefvanschie.inventoryframework.GuiItem;
import com.github.stefvanschie.inventoryframework.pane.OutlinePane;
import com.gmail.samueler53.fwhumanstratego.managers.ArenaManager;
import com.gmail.samueler53.fwhumanstratego.message.Message;
import com.gmail.samueler53.fwhumanstratego.objects.Arena;
import com.gmail.samueler53.fwhumanstratego.objects.Game;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.*;

public class GamesGui {
    private Gui mainGui;
    private final ArenaManager arenaManager = ArenaManager.getInstance();


    public GamesGui() {
        mainGui = prepareGui();
    }

    public Gui prepareGui() {
        mainGui = new Gui(1, "Games");
        mainGui.setOnGlobalClick(event -> event.setCancelled(true));

        OutlinePane background = new OutlinePane(0, 0, 9, 1);
        background.addItem(new GuiItem(new ItemStack(Material.BLACK_STAINED_GLASS_PANE)));
        background.setRepeat(true);
        mainGui.addPane(background);

        OutlinePane gamesPane = new OutlinePane(0, 0, 9, 1);
        mainGui.addPane(gamesPane);

        return mainGui;
    }

    public void createNewGame(Arena arena, Game game) {
        ItemStack itemStack = new ItemStack(Material.NETHERITE_BLOCK);
        ItemMeta itemMeta = itemStack.getItemMeta();
        if (itemMeta == null) return;
        itemMeta.setDisplayName(arena.getName());
        List<String> loreStrings = new ArrayList<>();
        loreStrings.add(game.getPlayersPlaying().size() + "/" + game.getNumberOfPlayers());
        itemMeta.setLore(loreStrings);
        itemStack.setItemMeta(itemMeta);
        addGame(itemStack, game);
    }

    private void addGame(ItemStack itemStack, Game game) {
        OutlinePane gamesPane = (OutlinePane) mainGui.getPanes().get(1);
        mainGui.setOnOutsideClick(event -> event.setCancelled(true));
        gamesPane.addItem(new GuiItem(itemStack, event -> addAPlayer(event.getWhoClicked().getUniqueId(), game)));
        mainGui.addPane(gamesPane);
        mainGui.update();
    }

    public void removeGame(Game game) {
        mainGui.getPanes().get(1).getItems().removeIf(GuiItem -> Objects.requireNonNull(GuiItem.getItem().getItemMeta()).getDisplayName().equalsIgnoreCase(game.getArena().getName()));
        mainGui.update();
    }

    public void modifyGame(Game game) {
        Collection<GuiItem> items = mainGui.getPanes().get(1).getItems();
        for (GuiItem guiItem : items) {
            if (Objects.requireNonNull(guiItem.getItem().getItemMeta()).getDisplayName().equalsIgnoreCase(game.getArena().getName())) {
                ItemMeta itemMeta = guiItem.getItem().getItemMeta();
                List<String> loreStrings = new ArrayList<>();
                loreStrings.add(game.getPlayersPlaying().size() + "/" + game.getNumberOfPlayers());
                itemMeta.setLore(loreStrings);
                guiItem.getItem().setItemMeta(itemMeta);
                mainGui.update();
            }
        }
    }

    public int getNumberOfGamesInTheGui() {
        return mainGui.getPanes().get(1).getItems().size();
    }

    public void show(UUID uuid) {
        Player player = Bukkit.getPlayer(uuid);
        if (player == null) return;
        mainGui.show(player);
    }

    private void addAPlayer(UUID uuid, Game game) {
        Player player = Bukkit.getPlayer(uuid);
        if (player == null) return;
        if (game.getNumberOfPlayers() != game.getPlayersPlaying().size()) {
            Message.GAME_JOIN.send(player);
            game.addAPlayer(uuid);
            game.getPlayersLocations().put(uuid, player.getLocation());
            arenaManager.teleportPlayerToLobby(uuid, game.getArena());
            modifyGame(game);
            game.clearPlayer(uuid);
            player.closeInventory();
        } else {
            Message.GAME_GAMEFULL.send(player);
        }
    }

}

File: src/main/java/com/gmail/samueler53/fwhumanstratego/gui/GamesGui.kt
package com.gmail.samueler53.fwhumanstratego.gui

import com.github.stefvanschie.inventoryframework.gui.GuiItem
import com.github.stefvanschie.inventoryframework.gui.type.ChestGui
import com.github.stefvanschie.inventoryframework.pane.OutlinePane
import com.gmail.samueler53.fwhumanstratego.FwHumanStratego
import com.gmail.samueler53.fwhumanstratego.objects.Arena
import com.gmail.samueler53.fwhumanstratego.objects.Game
import com.gmail.samueler53.fwhumanstratego.utils.editItemMeta
import com.gmail.samueler53.fwhumanstratego.utils.itemStack
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.meta.ItemMeta

@Suppress("Unused", "MemberVisibilityCanBePrivate")
class GamesGui {

    lateinit var gui: ChestGui

    lateinit var gamesPane: OutlinePane

    private val arenaGuiItems = mutableMapOf<String, GuiItem>()


    fun onGlobalClick(event: InventoryClickEvent) {
        event.isCancelled = true
    }

    fun onOutsideClick(event: InventoryClickEvent) {
        event.isCancelled = true
    }

    fun onGameCreated(arena: Arena, game: Game) {
        val itemStack = Material.NETHERITE_BLOCK.itemStack {
            editItemMeta {
                setDisplayName(arena.name)
                setGameInfoLoreStrings(game)
            }
        }
        val guiItem = GuiItem(itemStack) {
            game.onPlayerJoin(it.whoClicked as Player)
        }
        arenaGuiItems[arena.name] = guiItem
        gamesPane.addItem(guiItem)
        gui.update()
    }

    fun onGameRemoved(game: Game) {
        val guiItem = arenaGuiItems[game.arena.name] ?: return
        arenaGuiItems.remove(game.arena.name)
        gamesPane.removeItem(guiItem)
        gui.update()
    }

    fun updateGameInfo(game: Game) {
        arenaGuiItems[game.arena.name]?.item?.editItemMeta {
            setGameInfoLoreStrings(game)
        }
        gui.update()
    }

    private fun ItemMeta.setGameInfoLoreStrings(game: Game) {
        lore = listOf("${game.playersPlaying.size}/${game.numberOfPlayers}")
    }

    companion object {

        fun newInstance(): GamesGui {
            val res = FwHumanStratego.instance.getResource("games_gui.xml")!!
            return GamesGui().apply {
                gui = res.use { ChestGui.load(this, it)!! }
            }
        }
    }
}